### DISCUSSION TITLE
Saving HSSM models

### DISCUSSION THREAD

**Discussion by juneeybug** [lnccbrown/HSSM#371] [General]
Once I sample the model, by running model_1.sample, is there a way to save model_1 (the HSSM object)?

**Comment by digicosmos86** [comment]
Hi @juneeybug,

We currently don't support saving the model object. We do have plans to add this feature once both PyMC and Bambi standardize their workflow to save models. There might still be technical challenges to us, since we support many types of likelihoods and they are not easy to serialize.

One thing you can try is pickling the model object. I think it should work out of the box.

Thanks!
Paul

**Reply by shou-han** [reply]
I'm trying to pickle it, it says "can't pickle local object, 'make_distribution.<locals>$$MDistribution'. Can this be resolved? Thank you.

**Reply by digicosmos86** [reply]
Unfortunately, it seems that pickling might not work. As the error shows, we use a class called SSMDistribution which is returned from a function, which cannot be pickled. We recommend only saving the `InferenceData` object returned from the sampling as a netcdf file as described from the thread below. Let me know if you need more details

**Comment by AlexanderFengler** [comment]
Hi @juneeybug,

additional note on this. 

In principle it is never necessary to save the model object, you can simply reinstate and attach the traces from a previous run if you want to 'load' a model.

Best,
Alex

**Comment by XiaoyuZeng** [comment]
> you can simply reinstate and attach the traces from a previous run

Hi Alex,
**Could you please clarify how exactly can I reinstate and attach the traces?**

I also have the need to "load" my previous model, as it can take 1-2 days to finish a model and I don't want to re-run a model every time I need to do additional analysis. 

By now, I only figured out a way to save and load inference data using arviz (to_netcdf/from_netcdf). but I notice some of the function in hssm relied on hssm calss right?

Best,
Xiaoyu

**Comment by digicosmos86** [comment]
> > you can simply reinstate and attach the traces from a previous run
> 
> Hi Alex, **Could you please clarify how exactly can I reinstate and attach the traces?**
> 
> I also have the need to "load" my previous model, as it can take 1-2 days to finish a model and I don't want to re-run a model every time I need to do additional analysis.
> 
> By now, I only figured out a way to save and load inference data using arviz (to_netcdf/from_netcdf). but I notice some of the function in hssm relied on hssm calss right?
> 
> Best, Xiaoyu

Usually you only need the inference data object. If you want to "attach" the inference data to the model object, simply load the object with arviz, and then set `model.inference_obj = my_inference_data`. Note that this is more of a hack, since we don't recommend modifying the model object directly. However, not sure if there's any use case where this is necessary though. If there's any, please let us know, and we can make this a bit more explicit.
